### ISSUE TITLE
Message factory broken for files with message extensions in upb impl

### ISSUE THREAD

**Issue by alkasm** [protocolbuffers/protobuf#11863]
The Python `message_factory.GetMessages` and related functions/methods fail when registering an extension, since the upb library currently [does not implement `RegisterExtension`](https://github.com/protocolbuffers/upb/blob/b747edb830b0fab524e0063fb2e156c390405dfa/python/message.c#L1630-L1633) (see https://github.com/protocolbuffers/upb/issues/1163).

The Python and C++ message implementations work as expected.

**What version of protobuf and what language are you using?**
Version: protobuf==4.21.12 from PyPI
Language: Python

**What operating system (Linux, Windows, ...) and version?**

macOS 12.1

**What runtime / compiler are you using (e.g., python version or gcc version)**

Python library, upb message implementation.

**What did you do?**

Example proto file: `microext.proto`

```proto
syntax = "proto3";

package microext;

import "google/protobuf/descriptor.proto";

extend google.protobuf.MessageOptions {
  MicroExtension micro_extension = 50001;
}

message MicroExtension {
  string ext_a = 1;
  uint32 ext_b = 2;
}

message MicroTestMessage {
  string field_a = 1;
  bool field_b = 2;

  option (microext.micro_extension) = {
    ext_a: "extension-test"
    ext_b: 1234
  };
}
```

Generate a file descriptor set from `protoc`:
```sh
protoc -I ./ microext.proto --include_imports --descriptor_set_out ./microext.fds
```

Python code to read the file descriptor set and create dynamic message types: `test_msg_factory.py`

```python
from google.protobuf.descriptor_pb2 import FileDescriptorSet
from google.protobuf.message_factory import GetMessages

with open("microext.fds", "rb") as f:
    fds = FileDescriptorSet.FromString(f.read())

classes = GetMessages(fds.file)
MicroTestMessage = classes["microext.MicroTestMessage"]

msg = MicroTestMessage(field_a="test", field_b=True)
print(msg)
```

**What did you expect to see**

```
field_a: "test"
field_b: true
```

Note that when running with the Python message impl, or the C++ impl (properly installed), this ^ is the result I get from running e.g.
```
$ PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python python test_msg_factory.py
```

**What did you see instead?**

```
$ python test_msg_factory.py
Traceback (most recent call last):
  File "/Users/alkasm/prog/upb-proto-extensions/test_msg_factory.py", line 7, in <module>
    classes = GetMessages(fds.file)
  File "/Users/alkasm/prog/upb-proto-extensions/venv/lib/python3.9/site-packages/google/protobuf/message_factory.py", line 189, in GetMessages
    return _FACTORY.GetMessages([file_proto.name for file_proto in file_protos])
  File "/Users/alkasm/prog/upb-proto-extensions/venv/lib/python3.9/site-packages/google/protobuf/message_factory.py", line 158, in GetMessages
    extended_class.RegisterExtension(extension)
  File "/Users/alkasm/prog/upb-proto-extensions/venv/lib/python3.9/site-packages/google/protobuf/message.py", line 371, in RegisterExtension
    raise NotImplementedError
NotImplementedError
```

**Anything else we should know about your project / environment**

Note that there's a related change between main and the latest release on PyPI: https://github.com/protocolbuffers/protobuf/commit/c80e7efac72510a2bc3e9365520055f6d6656c1d, but I don't believe that affects this problem in either direction (neither causes or solves it).

**Comment by sashantp** [comment]
@esorot Is there anybody working on it ? Could you please inform us. Thanks!

**Comment by haberman** [comment]
@anandolee it looks like our unit tests don't cover this case. Could you add a unit test for this and fix upb accordingly?

**Comment by Krow10** [comment]
Very happy to see that an issue is already created for this !

I run into the exact same problem while trying to load protobuf classes from a descriptor set. A workaround that worked for me in the **4.21.12** version was to simply catch the exception in the `_FACTORY.GetMessages` function:
```python
from types import MethodType
from google.protobuf.descriptor_pb2 import FileDescriptorSet
from google.protobuf.message_factory import _FACTORY, GetMessages

def patch_get_messages(self, files):
    result = {}
    for file_name in files:
        file_desc = self.pool.FindFileByName(file_name)
        for desc in file_desc.message_types_by_name.values():
            result[desc.full_name] = self.GetPrototype(desc)

    for extension in file_desc.extensions_by_name.values():
        if extension.containing_type not in self._classes: #pylint: disable=protected-access
            self.GetPrototype(extension.containing_type)
        extended_class = self._classes[extension.containing_type] #pylint: disable=protected-access

        # === PATCH STARTS HERE ===

        try:
            extended_class.RegisterExtension(extension)
        except NotImplementedError:
            pass

        # === PATCH ENDS HERE ===

        if extension.message_type:
            self.GetPrototype(extension.message_type)
    return result

# Monkey-patching the `GetMessages` function to prevent `NotImplementedError` exceptions
_FACTORY.GetMessages = MethodType(patch_get_messages, _FACTORY)

# Use `GetMessages` normally
with open('protos.desc', 'rb') as proto_desc:
    descriptor_set = FileDescriptorSet.FromString(proto_desc.read())
print(GetMessages(descriptor_set.file))
```

That seemed to do the trick to be able to use `GetMessages`.

Upgrading to version **4.22.0** with the `_FACTORY` removed, we can apply the same patch to the [`GetMessageClassesForFiles`](https://github.com/protocolbuffers/protobuf/blob/main/python/google/protobuf/message_factory.py#L76) function:
```python
import google.protobuf.message_factory
from google.protobuf.descriptor_pb2 import FileDescriptorSet

def patch_get_messages_classes_for_files(files, pool):
    result = {}
    for file_name in files:
        file_desc = pool.FindFileByName(file_name)
        for desc in file_desc.message_types_by_name.values():
            result[desc.full_name] = google.protobuf.message_factory.GetMessageClass(desc)

    for extension in file_desc.extensions_by_name.values():
        extended_class = google.protobuf.message_factory.GetMessageClass(extension.containing_type)
        
        # === PATCH STARTS HERE ===

        try:
            extended_class.RegisterExtension(extension)
        except NotImplementedError:
            pass

        # === PATCH ENDS HERE ===

        if extension.message_type:
            google.protobuf.message_factory.GetMessageClass(extension.message_type)
    return result

# Monkey-patching the `GetMessages` function to prevent `NotImplementedError` exceptions
google.protobuf.message_factory.GetMessageClassesForFiles = patch_get_messages_classes_for_files

# Use `GetMessages` normally
with open('protos.desc', 'rb') as proto_desc:
    descriptor_set = FileDescriptorSet.FromString(proto_desc.read())
print(google.protobuf.message_factory.GetMessages(descriptor_set.file))
```
And it seems to hold fine when I test it alone. 

**However for my use case, this results in a `Segmentation fault` later on.** Not sure exactly why, I tried to profile it a little bit using `sys.settrace` and it seems to fail for subsequent calls to `MessageToJSON` using the loaded classes.

So to summarize, I'm not sure if the *segfaults* are related to this issue but fixing it would definitely eliminate the "monkey-patching business" out of the equation for me to be able to test it further. Until then, keeping to **4.21.12** seems to be the easiest option.

*PS: If you want to checkout the project I'm working on, it's available [here](https://github.com/pinax-network/substreams_firehose/tree/prepare-pypi).*

**Comment by cbeauchesne** [comment]
I've the same issue. Do you at which version the issue has been introduced in the first place?

**Comment by alkasm** [comment]
> I've the same issue. Do you at which version the issue has been introduced in the first place?

upb was introduced with the jump from protobuf 3.2 to 4.2. More here: https://protobuf.dev/news/2022-05-06/

**Comment by haberman** [comment]
Has anyone been able to create a repro for this yet?

**Comment by alkasm** [comment]
Note that support ends for python protobuf 3.20.x **this quarter**: https://protobuf.dev/support/version-support/#python

This is a breaking change that was introduced bumping from 3.20 to 4.21 which has not been remedied. Support for 3.20.x should be extended accordingly, or a fix should be included asap.

**Comment by haberman** [comment]
After more evaluation, we discovered that `RegisterExtension()` is no longer necessary unless you use the `Add*Descriptor()` functions, but those functions have been deprecated since August 2019: https://github.com/protocolbuffers/protobuf/commit/4f02f056b5cea99052bfdfb6698afe47a3cf2964#diff-c0e7fb116209fdc80b3d84c68595eb3158a72bb031a2e673fa04fbf88bf8db2b

If you use `AddSerializedFile()` instead, `RegisterExtension()` is not necessary.

Since upb does not support `Add*Descriptor()`, there is no need to add support for `RegisterExtension()`.  If you are using upb, you should be able to just remove any `RegisterExtension()` calls.

**Comment by anandolee** [comment]
A fix is on the way.

**Comment by alkasm** [comment]
Any updates here?

**Comment by gvashishtha** [comment]
@anandolee what's the status on the fix? This issue is blocking us from parsing protobufs correctly

**Comment by anandolee** [comment]
RegisterExtension() has been removed from message_factory.py. The issue should been fixed
